rohmu/object_storage/google.py

```diff
@@ -703,3 +701,92 @@ def _read_bytes(self, length: int, *, initial_data: Optional[bytes] = None) -> b
             return read_results[0]
         else:
             return b"".join(read_results)
+
+
+class MediaIoBaseDownload:
```

Comment:
prefer using the built-in one when byte_range is not provided.

Also iiuc, to get byte_range download just setting `_progress` would get use what we need ?
it is an internal attribute of the class but that is more preferrable than having a copy of that class here

https://googleapis.github.io/google-api-python-client/docs/epy/googleapiclient.http-pysrc.html#MediaIoBaseDownload

Reply:
plus a test, which creates the subclass and mocks out the retry_request method to checks if the correct headers are making it over, would be needed.

Reply:
@sjamgade I had already seen that it would be possible to hack around the implementation overriding the fields. 

I don't think `_progress` is enough though, there is also `_total_size` which needs to be worked around to report the correct progress and done status: https://github.com/googleapis/google-api-python-client/blob/main/googleapiclient/http.py#L765

We would have to have `_total_size` as some property/descriptor so that we can intercept here https://github.com/googleapis/google-api-python-client/blob/main/googleapiclient/http.py#L761 and replace it with the byte range size when needed... and that's really ugly.

Or we have to recompute all those fields from scratch in every call... and at that point you have rewritten half the class.

From my PV it's better to have our own implementation that works how we want, so we are sure that it won't ever change behaviour if we do a bugfix upgrade of the library.

If you want I can create an other branch with this more hackish solution and see how that works out, but my guess is that we don't save much code and make it more complicated and less stable due to possible internal changes in the library.

Reply:
@sjamgade  I have create a separate draft PR for what would look like messing with the internals in a subclass: https://github.com/aiven/rohmu/pull/110  . It is a bit shorter but to me this PR looks clearer.

If you think we should go with the other approach let me know and I will polish that one (I just hacked the implementation in the shortest time possible).

Reply:
true, this is cleaner given the details

Reply:
I'm going to clean up a few things and make sure we use our implementation only when the byte range is provided and then I will re-ask for review then.